src/jit/flowgraph.cpp

```diff
@@ -9751,123 +9751,162 @@ void                Compiler::fgCompactBlocks(BasicBlock* block, BasicBlock* bNe
     // TODO-CQ: This may be the wrong thing to do.  If we're compacting blocks, it's because a
     // control-flow choice was constant-folded away.  So probably phi's need to go away,
     // as well, in favor of one of the incoming branches.  Or at least be modified.
-    GenTreePtr blkNonPhi1   = block->FirstNonPhiDef();
-    GenTreePtr bNextNonPhi1 = bNext->FirstNonPhiDef();
-    GenTreePtr blkFirst     = block->firstStmt();
-    GenTreePtr bNextFirst   = bNext->firstStmt();
 
-    // Does the second have any phis?
-    if (bNextFirst != NULL && bNextFirst != bNextNonPhi1)
+    assert((block->IsLIR() == bNext->IsLIR()) || (block->IsLIR() && bNext->isEmpty()) || (block->isEmpty() && bNext->IsLIR()));
+    if (block->IsLIR() || bNext->IsLIR())
     {
-        GenTreePtr bNextLast = bNextFirst->gtPrev;
-        assert(bNextLast->gtNext == NULL);
+        LIR::Range blockRange = LIR::AsRange(block);
+        LIR::Range nextRange = LIR::AsRange(bNext);
 
-        // Does "blk" have phis?
-        if (blkNonPhi1 != blkFirst)
-        {
-            // Yes, has phis.
-            // Insert after the last phi of "block."
-            // First, bNextPhis after last phi of block.
-            GenTreePtr blkLastPhi;
-            if (blkNonPhi1 != NULL)
-            {
-                blkLastPhi = blkNonPhi1->gtPrev;
-            }
-            else
-            {
-                blkLastPhi = blkFirst->gtPrev;
-            }
+        GenTree* blockBegin = blockRange.Begin();
+        GenTree* blockFirstNonPhi = blockRange.FirstNonPhiNode();
 
-            blkLastPhi->gtNext = bNextFirst;
-            bNextFirst->gtPrev = blkLastPhi;
+        GenTree* nextBegin = nextRange.Begin();
+        GenTree* nextFirstNonPhi = nextRange.FirstNonPhiNode();
 
-            // Now, rest of "block" after last phi of "bNext".
-            GenTreePtr bNextLastPhi = NULL;
-            if (bNextNonPhi1 != NULL)
-            {
-                bNextLastPhi = bNextNonPhi1->gtPrev;
-            }
-            else
-            {
-                bNextLastPhi = bNextFirst->gtPrev;
-            }
+        GenTree* insertionPoint = blockFirstNonPhi != blockRange.End() ? blockFirstNonPhi : blockRange.EndExclusive();
 
```

Comment:
this seems wrong: if blockFirstNonPhi != nullptr, you'll insert AFTER the first nonphi. I think you need something like what I showed below. you should probably just use nullptr instead of blockRange.End()